File: 2021/Day1.hs
#! /usr/bin/env -S stack runhaskell
import Data.List

part1 :: [Int] -> Int
part1 [] = 0
part1 [_] = 0
part1 (x:y:xs) = (if y > x then 1 else 0) + part1 (y:xs)

part2 :: [Int] -> Int
part2 = part1 . slide
    where 
        slide [] = []
        slide [_] = []
        slide [_, _] = []
        slide (x:y:z:xs) = (x + y + z) : slide (y:z:xs)

main :: IO()
main = do
    f <- readFile("1.in")
    let input = map read $ lines f
    putStr "Solution for part 1: "
    putStrLn . show $ part1 input
    putStr "Solution for part 2: "
    putStrLn . show $ part2 input

-- Solution part 1: 1121
-- Solution part 2: 1065

File: 2021/Day1.kt
import java.io.File

typealias TInput1 = List<Int>
typealias TOutput1 = Int

private fun part1(input: TInput1): TOutput1 = input.windowed(2, 1).map { it[1] > it[0] }.count { it }

private fun part2(input: TInput1): TOutput1 = part1(input.windowed(3, 1).map { it.sum() })

private fun parse(raw_input: String): TInput1 = nums(raw_input)

fun main() {
    val rawInput = File("1.in").readText()
    val input = parse(rawInput)

    println("Solution for Part 1: ${part1(input)}")
    println("Solution for Part 2: ${part2(input)}")
}

// Solution part 1: 1121
// Solution part 2: 1065
